Polygon_mesh_processing/include/CGAL/Polygon_mesh_processing/internal/Isotropic_remeshing/remesh_impl.h

```diff
@@ -743,34 +742,55 @@ namespace internal {
           //do it before collapse is performed to be sure everything is valid
           if (!mesh_border_case)
             merge_status(en, s_epo, s_ep);
-          if (!mesh_border_case_opp)
+          if (!mesh_border_case_opp && s_ep_p!=PATCH_BORDER)
             merge_status(en_p, s_epo_p, s_ep_p);
 
           halfedge_and_opp_removed(he);
           if (!mesh_border_case)
             halfedge_and_opp_removed(prev(he, mesh_));
           if (!mesh_border_case_opp)
-            halfedge_and_opp_removed(prev(opposite(he, mesh_), mesh_));
-
-          //constrained case
-          bool constrained_case = is_constrained(va) || is_constrained(vb);
-          if (constrained_case)
           {
-            CGAL_assertion(is_constrained(va) ^ is_constrained(vb));//XOR
-            set_constrained(va, false);
-            set_constrained(vb, false);
+            if (s_ep_p!=PATCH_BORDER)
+              halfedge_and_opp_removed(prev(he_opp, mesh_));
+            else{
+              // swap edges so as to keep constained edges:
+              // replace en_p by epo_p and ep_p by eno_p
+              // (1) exchange next pointer
+              set_next(prev(epo_p, mesh_), en_p, mesh_);
+              set_next(en_p, next(epo_p, mesh_), mesh_);
+              set_next(prev(eno_p, mesh_), ep_p, mesh_);
+              set_next(ep_p, next(eno_p, mesh_), mesh_);
+              set_next(epo_p, eno_p, mesh_);
+              set_next(he_opp, epo_p, mesh_);
+              set_next(eno_p, he_opp, mesh_);
+              // (2) exchange vertex-halfedge pointer
+              set_target(ep_p, va, mesh_);
+              set_target(eno_p, vb, mesh_);
+              set_halfedge(va, ep_p, mesh_);
+              set_halfedge(vb, eno_p, mesh_);
+              // (3) exchange face-halfedge pointer
+              set_face(ep_p, face(eno_p, mesh_), mesh_);
+              set_face(en_p, face(epo_p, mesh_), mesh_);
+              set_face(epo_p, face(he_opp, mesh_), mesh_);
+              set_face(eno_p, face(he_opp, mesh_), mesh_);
+              set_halfedge(face(he_opp, mesh_), he_opp, mesh_);
+              if (face(ep_p, mesh_) != boost::graph_traits<PM>::null_face())
+                set_halfedge(face(ep_p, mesh_), ep_p, mesh_);
+              if (face(en_p, mesh_) != boost::graph_traits<PM>::null_face())
+                set_halfedge(face(en_p, mesh_), en_p, mesh_);
+              CGAL_assertion(mesh_.is_valid());
+            }
           }
 
           //perform collapse
-          Point target_point = get(vpmap_, vb);
+          CGAL_assertion(target(halfedge(e, mesh_), mesh_) == vb);
           vertex_descriptor vkept = CGAL::Euler::collapse_edge(e, mesh_);
-          put(vpmap_, vkept, target_point);
+          CGAL_assertion(mesh_.is_valid());
```

Comment:
this is not BGL

Reply:
yes `is_valid_polygon_mesh()` actually. Were you talking of something else?

Reply:
Ok, but then it should not be there... otherwise we may have issues with non-CGAL meshes.

Reply:
So what do we do ? we remove the assertion, making the function vulnerable ? Or we wait for the new mesh validity function to be integrated ?

Reply:
Why not use 
```
CGAL_assertion(is_valid(mesh_));
CGAL_assertion(is_triangle_mesh(mesh_));
```
as everywhere else in the file?

At least `is_valid(mesh_)` is in the `graph_traits_**.h` files